src/event/util/log/all_fields.rs

```diff
@@ -0,0 +1,170 @@
+use super::{Atom, Value};
+use serde::{Serialize, Serializer};
+use std::{
+    collections::{btree_map, BTreeMap},
+    iter, slice,
+};
+
+/// Iterates over all paths in form "a.b[0].c[1]" in alphabetical order
+/// and their corresponding values.
+pub fn all_fields<'a>(
+    fields: &'a BTreeMap<Atom, Value>,
+) -> impl Iterator<Item = (Atom, &'a Value)> + Serialize {
+    FieldsIter::new(fields)
+}
+
+#[derive(Clone)]
+enum LeafIter<'a> {
+    Map(btree_map::Iter<'a, Atom, Value>),
+    Array(iter::Enumerate<slice::Iter<'a, Value>>),
+}
+
+#[derive(Clone)]
+enum Node<'a> {
+    Key(&'a Atom),
+    Index(usize),
+}
```

Comment:
I'm trying to think if there's a more descriptive name we could use for this. `Node` makes me think of something like a linked list node that holds a value, whereas this is not quite that. Maybe something like `PathComponent`? That feels more precise, but it is unfortunately long.

Reply:
It was changed to `PathComponent`.